articles/tutorials/qft-tutorial.md

```diff
@@ -0,0 +1,649 @@
+
+#### Controlled operations
+
+Q# makes it extremely easy to condition the execution of an operation upon one or multiple control qubits.
+In general, we merely preface the call with `Controlled`, and the operation arguments change as:
+
+ `Op(<normal args>)` $\to$ `Controlled Op([<control qubits>], (<normal args>))`.
+
+Note that the control qubits must be provided as an array, even if it is a single qubit.
+
+After the `H`, we see that the next gates are the `R1` gates acting on the first qubit (and controlled by the second/third):
+
+<br/>
+<img src="./qft_firstqubit.PNG" alt="Circuit diagram for three qubit QFT through first qubit" width="310">
+
+We call these with
+
+```qsharp
+            Controlled R1([qs[1]], (PI()/2.0, qs[0]));
+            Controlled R1([qs[2]], (PI()/4.0, qs[0]));
+```
+
+Note that we use the [`PI()`](xref:microsoft.quantum.math.pi) function from the [`Microsoft.Quantum.Math`](xref:microsoft.quantum.math) namespace to define the rotations in terms of pi radians.
+Additionally, we divide by a `Double` (e.g. `2.0`) because dividing by an integer `2` would throw a type error. 
+
+> [!TIP]
+> `R1(π/2)` and `R1(π/4)` are equivalent to the `S` and `T` operations (also in `Microsoft.Quantum.Intrinsic`).
+
+
+After applying the relevant Hadamards and controlled rotations to the second and third qubits:
+
+```qsharp
+            //second qubit:
+            H(qs[1]);
+            Controlled R1([qs[2]], (PI()/2.0, qs[1]));
+
+            //third qubit:
+            H(qs[2]);
+```
+
+we need only apply a [`SWAP`](xref:microsoft.quantum.intrinsic.swap) gate to complete the circuit:
+
+```qsharp
+            SWAP(qs[2], qs[0]);
+```
+
+This is necessary because the nature of the quantum Fourier transform outputs the qubits in reverse order, so the swaps allow for seamless integration of the subroutine into larger algorithms.
+
+Hence we have finished writing the qubit-level operations of the quantum Fourier transform into our Q# operation:
+
+<img src="./qft_full.PNG" alt="Three qubit quantum Fourier transform circuit diagram" width="600">
+
+However, we can't call it a day just yet.
+Our qubits were in state $\ket{0}$ when we allocated them, and much like in life, in Q# we should leave things the same way we found them (or better!).
+
+### Deallocate qubits
+
+We call [`DumpMachine()`](xref:microsoft.quantum.diagnostics.dumpmachine) again to see the post-operation state, and finally apply [`ResetAll`](xref:microsoft.quantum.intrinsic.resetall) to the qubit register to reset our qubits to $\ket{0}$ before completing the operation:
+
+```qsharp
+            Message("After:");
+            DumpMachine();
+
+            ResetAll(qs);
+```
+
+Requiring that all deallocated qubits be explicitly set to $\ket{0}$ is a basic feature of Q#.
+If it is not performed at the end of a `using` allocation block, a runtime error will be thrown.
```

Comment:
Maybe we could explain here why? If I'm not wrong it is for other operations using the registers afterward to find them in the state 0. Since programs may involve several operations and qubits are a scarce resource, you want them ready to use anytime they're free. 
This makes me wonder, why is it chosen to leave qubits at "0" at the end of operation instead of at the beginning? My intuition tells me that it has to do with the error rate since applying extra measurements before the wanted result may increase the error rate. However, the same argument could be said about the measurements at the end. Maybe it is to also use qubits for other things apart from operations? Maybe @cgranade knows the answer.